ntoskrnl/ke/i386/kiinit.c

```diff
@@ -392,6 +340,92 @@ KiInitializePcr(IN ULONG ProcessorNumber,
     Pcr->PrcbData.MultiThreadProcessorSet = Pcr->PrcbData.SetMember;
 }
 
+static
+CODE_SEG("INIT")
+VOID
+KiVerifyCpuFeatures(PKPRCB Prcb)
+{
+    CPU_INFO CpuInfo;
+
+    // 1. Check CPUID support
+    ULONG EFlags = __readeflags();
+
+    /* XOR out the ID bit and update EFlags */
+    ULONG NewEFlags = EFlags ^ EFLAGS_ID;
+    __writeeflags(NewEFlags);
+
+    /* Get them back and see if they were modified */
+    NewEFlags = __readeflags();
+
+    if (NewEFlags == EFlags)
+    {
+        /* The modification did not work, so CPUID is not supported. */
+        KeBugCheckEx(UNSUPPORTED_PROCESSOR, 0x1, 0, 0, 0);
+    }
+    else
+    {
+        /* CPUID is supported. Set the ID Bit again. */
+        EFlags |= EFLAGS_ID;
+        __writeeflags(EFlags);
+    }
+
+    /* Peform CPUID 0 to see if CPUID 1 is supported */
+    KiCpuId(&CpuInfo, 0);
+    if (CpuInfo.Eax == 0)
+    {
+        // 0x1 - missing CPUID instruction
+        KeBugCheckEx(UNSUPPORTED_PROCESSOR, 0x1, 0, 0, 0);
+    }
+
+    // 2. Detect and set the CPU Type
+    KiSetProcessorType();
+
+    if (Prcb->CpuType == 3)
+        KeBugCheckEx(UNSUPPORTED_PROCESSOR, 0x386, 0, 0, 0);
+
+    // 3. Finally, obtain CPU features.
+    ULONG FeatureBits = KiGetFeatureBits();
+
+    // 4. Verify it supports everything we need.
+    if (!(FeatureBits & KF_RDTSC))
+    {
+        // 0x2 - missing CPUID features
+        // second paramenter - edx flag which is missing
+        KeBugCheckEx(UNSUPPORTED_PROCESSOR, 0x2, 0x00000010, 0, 0);
```

Comment:
Use the human-readable flag names (here and in the other places) instead of their hardcoded value.

Reply:
Oh please, we don't use them anywhere in the kernel

Reply:
I was thinking about using (for example) "KF_RDTSC" and co. ... unless there is not a one-to-one mapping between the actual CPU flag bit, and the "feature bits" stored in the eponymous name.

Reply:
Check `KeGetFeatureBits` https://github.com/reactos/reactos/blob/463d312169f6763fa51663e87d71e5abbd297331/ntoskrnl/ke/i386/cpu.c#L214 - we don't use any defines when accessing CPUID flags. That may be refactored, but not in this PR

Reply:
ok. Then I wonder, whether it's better to show up the actual CPU flag bit in the BSOD, or the NT `KF_xxx` flag instead 🤔

Reply:
I've choose CPUID flag because we don't have a feature bit for x87